source/matrix_free/vector_data_exchange.cc

```diff
@@ -316,6 +318,1057 @@ namespace internal
+        const ArrayView<const float> &             locally_owned_array,
+        const std::vector<ArrayView<const float>> &shared_arrays,
+        const ArrayView<float> &                   ghost_array,
+        const ArrayView<float> &                   temporary_storage,
+        std::vector<MPI_Request> &                 requests) const
+      {
+        import_from_ghosted_array_start_impl(vector_operation,
+                                             communication_channel,
+                                             locally_owned_array,
+                                             shared_arrays,
+                                             ghost_array,
+                                             temporary_storage,
+                                             requests);
+      }
+
+
+
+      void
+      Full::import_from_ghosted_array_finish(
+        const VectorOperation::values              vector_operation,
+        const ArrayView<float> &                   locally_owned_storage,
+        const std::vector<ArrayView<const float>> &shared_arrays,
+        const ArrayView<float> &                   ghost_array,
+        const ArrayView<const float> &             temporary_storage,
+        std::vector<MPI_Request> &                 requests) const
+      {
+        import_from_ghosted_array_finish_impl(vector_operation,
+                                              locally_owned_storage,
+                                              shared_arrays,
+                                              ghost_array,
+                                              temporary_storage,
+                                              requests);
+      }
+
+
+
+      template <typename Number>
+      void
+      Full::export_to_ghosted_array_start_impl(
+        const unsigned int                          communication_channel,
+        const ArrayView<const Number> &             data_this,
+        const std::vector<ArrayView<const Number>> &data_others,
+        const ArrayView<Number> &                   buffer,
+        const ArrayView<Number> &                   temporary_storage,
+        std::vector<MPI_Request> &                  requests) const
+      {
+#ifndef DEAL_II_WITH_MPI
+        Assert(false, ExcNeedsMPI());
+
+        (void)communication_channel;
+        (void)data_this;
+        (void)data_others;
+        (void)buffer;
+        (void)temporary_storage;
+        (void)requests;
+#else
+        (void)data_others;
+
+        requests.resize(sm_import_ranks.size() + sm_ghost_ranks.size() +
+                        ghost_targets_data.size() + import_targets_data.size());
+
+        int dummy;
+        // receive a signal that relevant sm neighbors are ready
+        for (unsigned int i = 0; i < sm_ghost_ranks.size(); i++)
+          MPI_Irecv(&dummy,
+                    0,
+                    MPI_INT,
+                    sm_ghost_ranks[i],
+                    communication_channel + 2,
```

Comment:
Can this number `communication_channel + 2` and `+ 3` further down mix up with any other non-blocking MPI communication elsewhere? In particular, I wonder how the `communication_channel` is incremented in the consumer code of `MatrixFree` and if different components of a block vector could clash. I believe they can't because the numbers are the same for all blocks of a vector.

Reply:
It looks the incrementation is two: 
https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/matrix_free.h#L3300

I need at least two tags. Should we increment then by 2?

Reply:
Since we are using `+0`, `+1`, `+2`, `+3` in this PR, I suggest to increase the counter there with `4 * component_in_block_vector`.